pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,389 @@
+		return err
+	}
+	// Start upgrade process for upgradesAvailable number of nodes
+	err = m.ProcessUpgradeRequiredNodes(ctx, currentState, upgradesAvailable)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(
+			err, "Failed to process nodes", "state", UpgradeStateUpgradeRequired)
+		return err
+	}
+	// Schedule nodes for drain
+	err = m.ProcessDrainNodes(ctx, currentState, upgradePolicy.DrainSpec)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to schedule nodes drain")
+		return err
+	}
+	err = m.ProcessPodRestartNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to schedule pods restart")
+		return err
+	}
+	err = m.ProcessDrainFailedNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes which failed to drain")
+		return err
+	}
+	err = m.ProcessUncordonRequiredNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to uncordon nodes")
+		return err
+	}
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, finished processing")
+	return nil
+}
+
+// ProcessDoneOrUnknownNodes iterates over UpgradeStateDone or UpgradeStateUnknown nodes and determines
+// whether each specific node should be in UpgradeStateUpgradeRequired or UpgradeStateDone state.
+func (m *ClusterUpgradeStateManager) ProcessDoneOrUnknownNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, nodeStateName string) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDoneOrUnknownNodes")
+
+	for _, nodeState := range currentClusterState.NodeStates[nodeStateName] {
+		podTemplateGeneration, err := utils.GetPodTemplateGeneration(nodeState.DriverPod, m.Log)
+		if err != nil {
+			m.Log.V(consts.LogLevelError).Error(
+				err, "Failed to get pod template generation", "pod", nodeState.DriverPod)
+			return err
+		}
+		if podTemplateGeneration != nodeState.DriverDaemonSet.GetGeneration() {
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateUpgradeRequired)
+			if err != nil {
+				m.Log.V(consts.LogLevelError).Error(
+					err, "Failed to change node upgrade state", "state", UpgradeStateUpgradeRequired)
+				return err
+			}
+			m.Log.V(consts.LogLevelInfo).Info("Node requires upgrade, changed its state to UpgradeRequired",
+				"node", nodeState.Node.Name)
+			continue
+		}
+
+		if nodeStateName == UpgradeStateUnknown {
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateDone)
+			if err != nil {
+				m.Log.V(consts.LogLevelError).Error(
+					err, "Failed to change node upgrade state", "state", UpgradeStateDone)
+				return err
+			}
+			m.Log.V(consts.LogLevelInfo).Info("Changed node state to UpgradeDOne",
+				"node", nodeState.Node.Name)
+		}
+		m.Log.V(consts.LogLevelInfo).Info("Node in UpgradeDone state, upgrade not required",
```

Comment:
can you change log level to debug here ?

also maybe add `continue` after L209 so you dont log the same info twice in case node state was unknown and its state was changed to done

Reply:
Changed the log level and added continue in the previous if statement